Add configurable retry and batch pause constants

pfefferle opened the pull request:
Added new configuration options to better manage traffic spikes when federating posts, allowing finer control over retry limits, delays, and batch pauses.

## Proposed changes:
- Add filters to change the default options

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [X] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Added new configuration options to better manage traffic spikes when federating posts, allowing finer control over retry limits, delays, and batch pauses.

</details>

pfefferle committed:
Add configurable retry and batch pause constants

Introduces ACTIVITYPUB_OUTBOX_RETRY_MAX_ATTEMPTS, ACTIVITYPUB_OUTBOX_RETRY_DELAY_UNIT, and ACTIVITYPUB_ASYNC_BATCH_PAUSE constants for improved control over retry logic and async batch scheduling. Updates Dispatcher and Scheduler classes to use these new constants for retries and batch pauses.

matticbot committed:
Add changelog

pfefferle committed:
Merge branch 'trunk' into customize-dispatcher

pfefferle committed:
Merge branch 'trunk' into customize-dispatcher

pfefferle commented:
@obenland I know that you are not a fan of consts, but that is how the rest works, so I tried to be consistent. Happy to change the whole system to use filters and/or options instead.

pfefferle committed:
Merge branch 'trunk' into customize-dispatcher

obenland commented:
Filters are probably the most flexible approach

pfefferle committed:
Refactor config to use getter methods and filters

Replaces direct usage of constants and static properties for batch size, retry attempts, delay unit, error codes, and async batch pause with getter methods that apply WordPress filters. This improves flexibility and allows these values to be customized via hooks. Removes related constants from constants.php.

obenland commented on includes/class-dispatcher.php:
```diff
@@ -19,21 +19,76 @@
+	 *
+	 * @return int The retry delay unit in seconds.
+	 */
+	public static function get_retry_delay_unit() {
```
```suggestion
	public static function get_retry_delay() {
```

obenland commented on includes/class-dispatcher.php:
```diff
@@ -19,21 +19,76 @@
+		 *
+		 * @param int $retry_delay_unit The retry delay unit in seconds. Default ACTIVITYPUB_OUTBOX_RETRY_DELAY_UNIT.
+		 */
+		return apply_filters( 'activitypub_dispatcher_retry_delay_unit', HOUR_IN_SECONDS );
```
```suggestion
		return apply_filters( 'activitypub_dispatcher_retry_delay', HOUR_IN_SECONDS );
```

obenland commented on includes/class-dispatcher.php:
```diff
@@ -95,13 +150,17 @@ public static function process_outbox( $id ) {
 	 * @return array|void The next batch of followers to process, or void if done.
 	 */
-	public static function send_to_followers( $outbox_item_id, $batch_size = ACTIVITYPUB_OUTBOX_PROCESSING_BATCH_SIZE, $offset = 0 ) {
+	public static function send_to_followers( $outbox_item_id, $batch_size = null, $offset = 0 ) {
```
Would `0` work instead of `null` to keep this an `int`? Or is there something else associated with `0`?

obenland commented on includes/class-scheduler.php:
```diff
@@ -32,6 +32,20 @@ class Scheduler {
+	 *
+	 * @return int The pause in seconds.
+	 */
+	public static function get_async_batch_pause() {
```
Pause is probably fine. Not sure if we should call it delay like in `get_retry_delay`?

pfefferle committed:
Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Refactor batch size and delay handling in dispatcher and scheduler

Updated Dispatcher to use ACTIVITYPUB_OUTBOX_PROCESSING_BATCH_SIZE as the default batch size. Renamed Scheduler's get_async_batch_pause() to get_retry_delay() and updated references to improve clarity and consistency in batch processing and scheduling.

pfefferle committed:
Use get_retry_delay instead of get_retry_delay_unit

Replaces the use of get_retry_delay_unit() with get_retry_delay() in the retry scheduling logic to ensure the correct delay value is applied for activity retries.

pfefferle commented:
Today was not mAI day :(
